lib/sshutils/sftp/sftp.go

```diff
@@ -322,95 +313,70 @@ func (c *Config) initFS(sshClient *ssh.Client, client *sftp.Client) error {
 		return nil
 	}
 
-	if c.getHomeDir == nil {
-		c.getHomeDir = func() (string, error) {
-			return getRemoteHomeDir(sshClient)
-		}
-	}
-
 	return trace.Wrap(c.expandPaths(srcOK, dstOK))
 }
 
 func (c *Config) expandPaths(srcIsRemote, dstIsRemote bool) (err error) {
-	srcHomeRetriever := getLocalHomeDir
 	if srcIsRemote {
-		srcHomeRetriever = c.getHomeDir
-	}
-	for i, srcPath := range c.srcPaths {
-		c.srcPaths[i], err = expandPath(srcPath, srcHomeRetriever)
-		if err != nil {
-			return trace.Wrap(err)
+		for i, srcPath := range c.srcPaths {
+			c.srcPaths[i], err = expandPath(srcPath)
+			if err != nil {
+				return trace.Wrap(err, "error expanding %q", srcPath)
+			}
 		}
 	}
 
-	dstHomeRetriever := getLocalHomeDir
 	if dstIsRemote {
-		dstHomeRetriever = c.getHomeDir
+		c.dstPath, err = expandPath(c.dstPath)
+		if err != nil {
+			return trace.Wrap(err, "error expanding %q", c.dstPath)
+		}
 	}
-	c.dstPath, err = expandPath(c.dstPath, dstHomeRetriever)
 
-	return trace.Wrap(err)
-}
-
-func getLocalHomeDir() (string, error) {
-	u, err := user.Current()
-	if err != nil {
-		return "", trace.Wrap(err)
-	}
-	return u.HomeDir, nil
+	return nil
 }
 
-func expandPath(pathStr string, getHomeDir homeDirRetriever) (string, error) {
-	if !needsExpansion(pathStr) {
+func expandPath(pathStr string) (string, error) {
+	pfxLen, ok := homeDirPrefixLen(pathStr)
+	if !ok {
 		return pathStr, nil
 	}
 
-	homeDir, err := getHomeDir()
-	if err != nil {
-		return "", trace.Wrap(err)
+	// Removing the home dir prefix would mean returning an empty string,
+	// which is supported by SFTP but won't be as clear in logs or audit
+	// events. Since the SFTP server will be rooted at the user's home
+	// directory, "." and "" are equivalent in this context.
```

Comment:
Should we handle `""` in the same way as `"~"` then? Can we ever get an empty `pathStr` argument here?

Reply:
We can't get an empty string here, `CreateUploadConfig` and `CreateDownloadConfig` ensure that the source and destination paths are not empty, and the `Config` struct's important fields are not exported meaning users can't construct it themselves